enhancements/installer/byo-custom-dns.md

```diff
@@ -0,0 +1,558 @@
+
+12. During cluster deletion, OpenShift is responsible only for deleting the
+Ingress LB and the customer is responsible for deleting the LBs for API and
+API-Int. The customer would also be responsible for removing entries from
+their custom DNS.
+
+#### Variation [optional]
+
+ARO was able to solve this problem in a neat way by making use of the
+[`LoadBalancerIP`] (https://github.com/kubernetes/api/blob/cdff1d4efea5d7ddc52c4085f82748c5f3e5cc8e/core/v1/types.go#L4721) field of the LoadBalancer service manifest.
+
+By using a pre-defined static IP as input to the LoadBalancer Service with the
+same name and annotations as one created by the cluster-ingress-operator, ARO
+is able to create an Ingress LB during the bootstrap phase. The non-cloud DNS
+could be pre-configured to resolve *.apps service to this pre-defined static
+IP.
+
+This would work well except for the following 2 reasons:
+1. AWS does not refer to the `LoadBalancerIP` input value before it creates an
+Ingress LB. So, the above solution only would work for Azure and GCP and not
+for AWS.
+
+2. Based on [K8s API documentation] (https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/#servicespec-v1-core), the `LoadBalancerIP` field would soon be deprecated.
+
+### API Extensions
+
+1. Install config is augmented to allow the customer to specify if a custom
+DNS solution is being used. In addition, the customer would be able to indicate
+the LB type to be used for API and API-Int traffic.
+
+```yaml
+---
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  annotations:
+    controller-gen.kubebuilder.io/version: v0.9.2
+  creationTimestamp: null
+  name: installconfigs.install.openshift.io
+spec:
+  group: install.openshift.io
+  names:
+    kind: InstallConfig
+    listKind: InstallConfigList
+    plural: installconfigs
+    singular: installconfig
+  scope: Namespaced
+  versions:
+  - name: v1
+    schema:
+<snip>
+                    aws:
+                      description: AWS is the configuration used when installing on
+                        AWS.
+                      properties:
+                        userConfiguredDNSLB:
+                          description: "UserConfiguredDNSLB contains all the API and API-Int
+                            LB information. \n This field is used to Enable the use of a
+                            custom DNS solution when the DNS provided by the underlying
+                            cloud platform cannot be used. When Enabled, the user can provide
+                            information about user created API and API-Int LBs using this
+                            field."
+                          properties:
+                            apiIntLBName:
+                              description: ApiIntLBName is the name of the API-Int NLB created
+                                by the user.
+                              type: string
+                            apiLBName:
```

Comment:
If user is installing a private cluster, this property would be optional, right?

Reply:
Or the feature is only applicable for a public cluster.

Reply:
This enhancement is talking about adding this functionality to AWS. The plan is to extend this capability to Azure and GCP in future releases.

Reply:
@jianlinliu Hopefully your question is answered in the latest update.

Reply:
Per my understanding, if I am installing a private cluster cluster on aws, byo dns is allowed, so I can set the install-config.yaml like the following:
```
platform:
  aws:
    userConfiguredDNSLB:
      apiIntLBName: <my-api-int-lb-lb-name>
      userDNS: Enabled
publish: Internal
```
So this is a validate test scenario, right?

Reply:
It is. And you would also have to provide the apiLBName.

Reply:
@sadasu but a private cluster never need a public api lb, in that case, how to move on?